Ports Polaris integrated electronic circuits

PsiOmegaDelta opened the pull request:
Might lead to interesting constructs in the future.
Have Bay-tweaked the code and dumped the gun firing circuit (because the code it led to was copy-paste work that made me feel less than good) and the Exonet circuit (because we don't currently have it).

During tweaking several potentially subtle bugs were discovered, local testing and unit tests forthcoming.

PsiOmegaDelta committed:
Ports Polaris integrated electronic circuits, for potentially interesting assemblies.

PsiOmegaDelta committed:
Bay tweaks the integrated circuits.

BlueNexus commented:
I already have some uses for this in mind. Looking forward to this.

PsiOmegaDelta commented:
Some quick testing reveals that the port appears to have been successful and that the integrated circuits work.

Only some minor issues with spots of code duplication, the UI is useful but feels a little awkward, and some operations with the tools are annoying (for example, you can cancel out wiring components together but that also toggles the mode from connecting to disconnecting).

Raptor1628 commented:
Can those of us who don't know what this is get a tl;dr?

PsiOmegaDelta commented:
Well, the tl;dr is 
![A typical breadboard](https://upload.wikimedia.org/wikipedia/commons/4/47/Breadboard_complex.jpg)

PsiOmegaDelta commented:
Slightly less tl;dr:
It's a system with what's basically a glorified breadboard (assemblies) and with circuits that you can configure and connect as desired.
Examples of circuits include timers, beepers, health scanners, banks of memory, logic units (AND, OR, NOR, etc.), buttons and so forth.

They are not there quite yet but once fully realized these assemblies could do amazing things, like fully configurable portable turrets.

A simpler example would be the laser emitter. With a button, 1 bit memory chip, NOR gate, timer and a well-done gun firing mechanism an emitter could be turned into little more than a frame that holds a weapon and said assembly.
With a press of the button the memory would get the inverse of its current value and if positive the timer would engage and every pulse it'd trigger the firing mechanism. In turn, the laser could be a variant that becomes stronger with cooldown and allow picking either a slow timer for few, strong shots or a fast one for a weaker barrage.
Then we could write an id reader circuit used to enable/disable the button which would result in the same operation lock behavior we have today.

Another even simpler example would be a C4 slab. Imagine being able to connect a timer assembly to the already existing go-boom-wire and when a clock reaches a certain time it gives it a pulse and.. well.

Basically, less hard-coded stuff more emergent game play.

PsiOmegaDelta committed:
Ports https://github.com/PolarisSS13/Polaris/pull/2504 with Bay tweaks.

BlueNexus commented:
In regards to your emitter example, how would that work without the gun firing circuit?

BlueNexus commented:
Also, another question - Will it be possible to find and use these components and assemblies in-game? People could make all kinds of things with these, and I can see it being very useful.

BlueNexus commented:
So, how are people going to obtain these IC? I can see these being very useful and interesting to people who know how to use them. Perhaps make them an R&D thing, or make the circuit kits available somehow? I could write a wiki page for them, if you do.

Raptor1628 commented:
Why not just call them breadboards then? Easier to understand considering every circuit is electronic, and can be integrated with SOMETHING.

BlueNexus commented:
Because that wouldn't describe all the components, just the assembly itself, @Raptor1628 
Also all of their assemblies are metal, with the circuits and stuff going inside, so they aren't just breadboards at all.

Just my thoughts on the matter, not my PR.

BlueNexus commented:
Bugs I've found:
- [x] Input devices can only be used while the assembly is crowbarred open. This includes buttons, keypads, and number pads. If this is intended, then that is very strange, and makes no sense.
- [x] Converters have no output pin
- [x] Component limit seems to be one too low - When trying to insert a component into a part with 9/10 component limit, it says it is over the limit.

PsiOmegaDelta committed:
As the ticker process() proc has been rewritten to use proper process ticks ticks_to_pulse vars have been adjusted accordingly.

Also gives the ticker input a purpose.

PsiOmegaDelta committed:
Variety of tweaks.

write_data_to_pin now correctly verifies the /weakref type.
Moves the component removal into the assembly UI instead of being an attackby operation.
Wirer, debugger, screwdriver now all open the assembly interface.
Screen outputs should now always be available, whether or not the assembly is open.
Should now be possible to fill an assembly to capacity, as opposed to 'almost'.
Converters should now properly setup outputs/activators.
Beeper frequency shifting is now consistent.
Some changes in wording, the big assembly toolkit now contains 4 of each component as originally intended.
The ticker handles data changes in a sturdier way now.
Logic circuits should be a bit sturdier.

PsiOmegaDelta commented:
Currently assemblies have the appearance of containers of various sorts, computers, shell of a drone, etc. and they can also be shut close to prevent further (non-violent) changes, and as such they are a little more than merely breadboards.

mwerezak commented:
Really cool idea.

It sounds like the utility of this system will increase greatly with the variety of objects that can be used with IC assemblies, so it might be productive to think about (or enlist people to brainstorm) different ways that existing items could be tied into the system.

@Raptor1628 IRL integrated circuits are [these things](https://upload.wikimedia.org/wikipedia/commons/8/80/Three_IC_circuit_chips.JPG) that can be placed onto a breadboard (or soldered onto a circuit board for a more permanent construction).

BlueNexus commented:
@Mwerezak It's very easy to tie into this system, using signalers. You can attach a signaler to lots of things like proxy sensors etc, and just have them send a signal which an integrated signaler can receive. Other ways to tie into it would definitely be nice, So I agree.

An example:
http://www.bluenexus.xyz/storage/i/2016-09-21_18-40-16.gif

BlueNexus commented:
If I may make a request, we could really use an NTNet equivalent of the EP2N circuit on polaris. The inter-device communication capability it adds would be extremely useful, since all we can currently use to do that is signalers, which can't transfer data at all, just a signal.

PsiOmegaDelta commented:
Probably worthwhile but too big to include in this PR, even as a near-empty shell.
Other than that, since you've been testing quite a bit @BlueNexus, noted any issues since the last commit of this PR?

BlueNexus commented:
I remember encountering some issues where NOT gates would stop receiving new values on their input pin, and manually entering a new value then pulsing the compare pin did nothing. In addition, beeper and Beepsky circuits don't seem to work at all, they make no sound. 

There's still a lot to be done though, like making them obtainable IC. Perhaps put circuit kits in engineering vending machines? Or do you want to save that for a second PR?

PsiOmegaDelta commented:
Fixed a bug with speakers. Tried to make logic circuits a bit sturdier.
Also, I'd rather separate implementation and the providing.

Neerti committed:
Fixes some problems from port, mainly ref stuff.

PsiOmegaDelta committed:
Defines data/data_to_write as weakrefs, reducing the number of necessary casts.

Neerti committed:
Adds More Circuits

Drone assembly can be made in RnD now.
Adds injector component, which does what it implies, and injects a reagent into someone or something nearby.
Adds reagent pump, similar to the above but meant for moving reagents around different containers, inside or outside of the machine.
Adds adjacent locator, which can find specific objects next to the machine.
Adds two flavors of the reagent storage component, which does what it says.  The second variant is a cryo-beaker version.

# Conflicts:
#	code/modules/research/designs.dm

PsiOmegaDelta committed:
Additional Bay tweaks.

Can now define pin startup values.
Broke out the injection checks into a general proc.

PsiOmegaDelta commented:
Ported https://github.com/PolarisSS13/Polaris/pull/2541.
Tweaked it by making it possible to define pin startup values, and broke out the injection checks into a general proc.

BlueNexus commented:
An odd bug I found: http://i.imgur.com/XEMP8In.gifv

PsiOmegaDelta commented:
Not sure what is going on. Would it be possible to explain in text:
* How to arrange the setup.
* How to act on the setup.
* How to assert the result.

BlueNexus commented:
Here's a description of a bug I found, written using Ginja's version of ICN.

```
subtraction circuit (sub)

sub A = {number}
sub B = {number}
sub compute *
//Both numbers can be set to anything you like, the output will always be zero.
```

I'll post the ICN for another, more serious bug shortly

PurpleMartinJCK commented:
Bug that I found: when using a memory chip, it can accept values perfectly well on the left/input side. On the right/output side the set button correctly transferred the values from the left to the right but it didn't push those values out the wire to the AND gates I had hooked up. They only got a null value.

BlueNexus commented:
I was thinking, and a text2ascii and ascii2text circuit (which are both supported by BYOND) would make transmission of text possible, as a replacement for EPV2. I already have a method of transferring numbers between devices, based on the delay between two signals.

PsiOmegaDelta committed:
Additions and fixes.

Subtraction should again work.
Memory circuits should now push their output after update.
Adds string2ascii and ascii2string converters.

PsiOmegaDelta commented:
Subtraction should again work.
Memory circuits should now push their output after update.
Adds string2ascii and ascii2string converters.

PsiOmegaDelta committed:
Adds a grenade assembly.

Possible to connect any grenade to the assembly.
Possible to set a custom detonation time between 1 to 12 seconds.

PsiOmegaDelta commented:
Adds a grenade assembly.

Possible to connect any grenade to the assembly.
Possible to set a custom detonation time between 1 to 12 seconds.

I experimented with having the grenades added as overlays to the circuit. Worked sort of-ish well, the grenades tended to be off-center and wouldn't show up at all while the circuit was held in the hand.

PurpleMartinJCK commented on code/modules/integrated_electronics/memory.dm:
```diff
@@ -0,0 +1,154 @@
+		var/datum/integrated_io/I = inputs[i]
+		var/datum/integrated_io/O = outputs[i]
+		O.data = I.data
+	push_data()
```
Should this be 'O.push_data()' instead of 'push_data()' ?

PsiOmegaDelta replied:
push_data() pushes updates for all outputs at once.
In case it ever matters this method ensures that all outputs are updated before informing connected pins of said update.

PurpleMartinJCK replied:
Ah ok. I was only asking as the memory chips weren't working when I tested it on Tuesday. It wasn't sending its output to the input of an and gate so the and gate was stuck on null for that input pin.

BlueNexus replied:
Psi updated it to hopefully fix the memory circuits, @FTangSteve

comma commented on code/modules/integrated_electronics/input_output.dm:
```diff
@@ -0,0 +1,404 @@
+	if(istext(ID.data) && isnum(vol.data) && isnum(frequency.data))
+		var/selected_sound = sounds[ID.data]
+		if(!selected_sound)
+			world << "No sound"
```
Debug print?

PsiOmegaDelta replied:
Now removed.

PsiOmegaDelta committed:
Adds filtering circuits.

Can now filter between various types of ref input. Will be useful together with future sensor circuits which will pickup refs from nearby targets and send them to other components.

PsiOmegaDelta commented:
Can now filter various types of ref input. Will be useful together with future sensor circuits which will pickup refs from nearby targets and send them to other components.
